### PR TITLE
improve caching method

### PR DESCRIPTION
**Pull request by tharindu1st**

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/WebsocketInboundHandler.java [hunk 1]
```diff
@@ -259,20 +263,19 @@ private boolean validateOAuthHeader(FullHttpRequest req) throws APISecurityExcep
...
                 //Initial guess of a JWT token using the presence of a DOT.
 
-                SignedJWT signedJWT = null;
+                SignedJWTInfo signedJWTInfo = null;
                 String keyManager = null;
                 if (StringUtils.isNotEmpty(apiKey) && apiKey.contains(APIConstants.DOT)) {
                     try {
                         // Check if the header part is decoded
-                        Base64.getUrlDecoder().decode(apiKey.split("\\.")[0]);
                         if (StringUtils.countMatches(apiKey, APIConstants.DOT) != 2) {
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/WebsocketInboundHandler.java:271]: we don't need this as well right? I think that will be handled in SignedJWT.parse()

Comment (tharindu1st) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/WebsocketInboundHandler.java:271]: its already deleted

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/WebsocketInboundHandler.java [hunk 2]
```diff
@@ -585,4 +588,28 @@ private void removeTokenFromQuery(Map<String, List<String>> parameters) {
...
+        String signature = accessToken.split("\\.")[2];
+        SignedJWTInfo signedJWTInfo = new SignedJWTInfo();
+        Cache gatewaySignedJWTParseCache = CacheProvider.getGatewaySignedJWTParseCache();
+        if (gatewaySignedJWTParseCache != null) {
+            Object cachedEntry = gatewaySignedJWTParseCache.get(signature);
+            if (cachedEntry != null) {
+                signedJWTInfo = (SignedJWTInfo) cachedEntry;
+            } else {
+                SignedJWT signedJWT = SignedJWT.parse(accessToken);
+                signedJWTInfo.setSignedJWT(signedJWT);
```
Comment (jaadds) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/WebsocketInboundHandler.java:602]: This part should go into SignedJWTInfo. So SignedJWTInfo has a constructor which accepts a string constructor. Inside it the the signedJWT is created. Then while calling to SignedJWTInfo.getClaimSet method, the getClaimSet method of the signedJWT is called and resulting claims are cached at the SignedJWTInfo layer. getClaimSet is not called prematurely but only when it's needed. And the ClaimSet is saved for subsequent calls.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/WebsocketInboundHandler.java [hunk 3]
```diff
@@ -585,4 +588,28 @@ private void removeTokenFromQuery(Map<String, List<String>> parameters) {
...
+                signedJWTInfo.setJwtClaimsSet(signedJWT.getJWTClaimsSet());
+                signedJWTInfo.setToken(accessToken);
+                gatewaySignedJWTParseCache.put(signature, signedJWTInfo);
+            }
+        } else {
+            SignedJWT signedJWT = SignedJWT.parse(accessToken);
+            signedJWTInfo.setSignedJWT(signedJWT);
```
Comment (jaadds) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/WebsocketInboundHandler.java:609]: Repeating block. Better move this to SignedJWTInfo class as mentioned in the previous comment.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/jwt/SignedJWTInfo.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+        return signedJWT;
+    }
+
+    public void setSignedJWT(SignedJWT signedJWT) {
+
+        this.signedJWT = signedJWT;
+    }
+
+    public JWTClaimsSet getJwtClaimsSet() {
```
Comment (jaadds) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/jwt/SignedJWTInfo.java:45]: Change this to check for stored jwtClaimsSet and compute if absent.

**Pull request merged**